include/gridtools/fn/backend2/gpu.hpp

```diff
@@ -0,0 +1,165 @@
+#include "../../sid/contiguous.hpp"
+#include "../../sid/multi_shift.hpp"
+#include "./common.hpp"
+
+namespace gridtools::fn::backend {
+    namespace gpu_impl_ {
+        template <class BlockSizes>
+        struct gpu {
+            using block_sizes_t = BlockSizes;
+        };
+
+        template <class BlockSizes, class Dims, int I>
+        using block_size_at_dim = meta::second<meta::mp_find<BlockSizes, meta::at_c<Dims, I>>>;
+
+        template <class BlockSizes, class Sizes>
+        GT_FUNCTION_DEVICE auto global_thread_index() {
+            using ndims_t = meta::length<Sizes>;
+            using keys_t = meta::take_c<std::min(3, (int)ndims_t::value), get_keys<Sizes>>;
+            if constexpr (ndims_t::value == 0) {
+                return hymap::keys<>::values<>();
+            } else if constexpr (ndims_t::value == 1) {
+                using block_dim_x = block_size_at_dim<BlockSizes, keys_t, 0>;
+                using values_t = typename keys_t::template values<int>;
+                return values_t(blockIdx.x * block_dim_x::value + threadIdx.x);
+            } else if constexpr (ndims_t::value == 2) {
+                using block_dim_x = block_size_at_dim<BlockSizes, keys_t, 0>;
+                using block_dim_y = block_size_at_dim<BlockSizes, keys_t, 1>;
+                using values_t = typename keys_t::template values<int, int>;
+                return values_t(
+                    blockIdx.x * block_dim_x::value + threadIdx.x, blockIdx.y * block_dim_y::value + threadIdx.y);
+            } else {
+                using block_dim_x = block_size_at_dim<BlockSizes, keys_t, 0>;
+                using block_dim_y = block_size_at_dim<BlockSizes, keys_t, 1>;
+                using block_dim_z = block_size_at_dim<BlockSizes, keys_t, 2>;
+                using values_t = typename keys_t::template values<int, int, int>;
+                return values_t(blockIdx.x * block_dim_x::value + threadIdx.x,
+                    blockIdx.y * block_dim_y::value + threadIdx.y,
+                    blockIdx.z * block_dim_z::value + threadIdx.z);
+            }
+        }
+
+        template <class Index, class Sizes>
+        GT_FUNCTION_DEVICE bool in_domain(Index const &index, Sizes const &sizes) {
+            return tuple_util::device::all_of(std::less(), index, sizes);
+        }
+
+        template <class BlockSizes, class Sizes, class PtrHolder, class Strides, class Fun>
+        __global__ void kernel(Sizes sizes, PtrHolder ptr_holder, Strides strides, Fun fun) {
+            auto thread_idx = global_thread_index<BlockSizes, Sizes>();
+            if (!in_domain(thread_idx, sizes))
+                return;
+            auto ptr = ptr_holder();
+            sid::multi_shift(ptr, strides, thread_idx);
+            if constexpr (meta::length<Sizes>::value <= 3) {
+                fun(ptr, strides);
+            } else {
+                common::make_loops(tuple_util::device::drop_front<3>(sizes))(wstd::move(fun))(ptr, strides);
```

Comment:
`drop_front` here will not work unfortunately. the right thing to do is to get keys from `sizes`; meta::drop_front` them and pass them as an additional `make_loops` parameter.